### PR TITLE
Mei/add col order

### PR DESCRIPTION
**Pull request by BAIMEI1**
the problem the pr solve:
- add colOrder handle logic. if our user does hope that content in a row in order, we can provide this feature.
- make 2D method more reasonable, the user does not have to use a column which is named "row".
users just need tell us the rowOrderCol name then we will handle it.

### CODE REVIEW

File: src/main/scala/io/github/fucusy/Viz.scala [hunk 1]
```diff
@@ -35,60 +36,104 @@ object Viz {
...
-            item.toString
-          }
-        }
-      }
 
-    val actualLimitShowNumber = if (limitShowNumber == -1) {
-      collectDF(0).size
-    } else {
-      limitShowNumber
+  private def addRowTitle(rowTitle: Option[String])(df: DataFrame): DataFrame = {
```
Comment (fucusy) [inline src/main/scala/io/github/fucusy/Viz.scala]: A few lines code doesn't worth creating a new function

File: src/main/scala/io/github/fucusy/Viz.scala [hunk 2]
```diff
@@ -35,60 +36,104 @@ object Viz {
...
+    * @return
+    */
 
   def dataframe2html2D(df: DataFrame,
-                       limitShowNumber: Int = -1): String = {
-    require(df.columns.contains("row") && df.columns.contains("title"))
-    val rowNum = df.select("row").distinct().count().toInt
-    val tables = (1 to rowNum)
+                       rowOrderCol: String,
+                       colOrderCol: Option[String] = None,
```
Comment (fucusy) [inline src/main/scala/io/github/fucusy/Viz.scala]: To make the logic clear, let's required to have row, col, title columns. 
For our case, we have row, video_position, container_slug.  

We can make our API more flexible when necessary.

File: src/main/scala/io/github/fucusy/Viz.scala [hunk 3]
```diff
@@ -35,60 +36,104 @@ object Viz {
...
+  private def addColOrder(colOrder: Option[String])(df: DataFrame): DataFrame = {
+    if(colOrder != None) {
+      df
+    }else{
+      df.withColumn("col_order", F.row_number.over(Window.orderBy(F.col(df.columns(0)).desc)))
+    }
+  }
+
+  def dataframe2data(df: DataFrame,
+                     colOrderCol: Option[String] = None,
```
Comment (fucusy) [inline src/main/scala/io/github/fucusy/Viz.scala]: We will assume the data frame is ordered before using the function dataframe2data.

Comment (fucusy) [reply src/main/scala/io/github/fucusy/Viz.scala]: So, we don't need to update this function

**Review by fucusy — CHANGES_REQUESTED** [review]

File: src/main/scala/io/github/fucusy/Viz.scala (hunk 3 above)
Comment (BAIMEI1) [reply src/main/scala/io/github/fucusy/Viz.scala]: no, dataframe in spark are not in order!! when I try to do 2d method, the video_position did not in order at all. but in fact we need it to keep order. https://dbc-5f2acc18-b29d.cloud.databricks.com/?o=6211501775943445#notebook/2550389456839224/command/2550389456839225
image

File: src/main/scala/io/github/fucusy/Viz.scala (hunk 2 above)
Comment (BAIMEI1) [reply src/main/scala/io/github/fucusy/Viz.scala]: I think for 2D method we require rowOrderCol is enough, colOrderCol and rowrTitleCol can be empty logically. if users want to show these two fields. they can do that. as long as we can support more, why we do not support it?

File: src/main/scala/io/github/fucusy/Viz.scala (hunk 1 above)
Comment (BAIMEI1) [reply src/main/scala/io/github/fucusy/Viz.scala]: I just want to separate it from the main logic. it's a relatively independent logic. I think we should create a method by function but not how many lines do they have.

File: src/main/scala/io/github/fucusy/Viz.scala [hunk 4]
```diff
@@ -41,26 +42,45 @@ object Viz {
...
-    val tables = (1 to rowNum)
+                       rowOrderCol: String,
+                       colOrderCol: String,
+                       rowTitleCol: Option[String] = None,
+                       limitShowNumber: Int = -1
+                       ): String = {
+    require(df.columns.contains(rowOrderCol) && df.columns.contains(colOrderCol))
+    val addRowTitleDF = if(rowTitleCol != None) {df}
+    else{
+      df.withColumn("row_title", F.row_number.over(Window.orderBy(F.col(df.columns(0)).desc)))
```
Comment (fucusy) [inline src/main/scala/io/github/fucusy/Viz.scala]: are we setting the default title column to some row number. df.columns(0) could be some unknown column.

Can we do this to simplify the logic:
```
df.withColumn("row_title", F.col(rowTitleCol.getOrElse(rowOrderCol))
```

File: src/main/scala/io/github/fucusy/Viz.scala [hunk 5]
```diff
@@ -41,26 +42,45 @@ object Viz {
...
+    }
+
+    val rowTitleColumn: Column = rowTitleCol match {
+      case Some(rowTitle) => df.col(rowTitle)
+      case None      => df.col("row_title")
+    }
+    val rowNumList = addRowTitleDF.select(rowOrderCol)
+      .distinct()
+      .collect
+      .map(_.getAs[Int](rowOrderCol))
```
Comment (fucusy) [inline src/main/scala/io/github/fucusy/Viz.scala:72]: we only have one column here, 
```
.map(_.toString.toInt)
```
should work

**Review by fucusy — CHANGES_REQUESTED** [review]
Leave two comments, Others LGTM

File: src/main/scala/io/github/fucusy/Viz.scala (hunk 4 above)
Comment (BAIMEI1) [reply src/main/scala/io/github/fucusy/Viz.scala]: got, changed

File: src/main/scala/io/github/fucusy/Viz.scala (hunk 5 above)
Comment (BAIMEI1) [reply src/main/scala/io/github/fucusy/Viz.scala:72]: no, it did not work
image

File: src/main/scala/io/github/fucusy/Viz.scala [hunk 6]
```diff
@@ -41,26 +42,45 @@ object Viz {
...
+                       rowOrderCol: String,
+                       colOrderCol: String,
+                       rowTitleCol: Option[String] = None,
+                       limitShowNumber: Int = -1
+                       ): String = {
+    require(df.columns.contains(rowOrderCol) && df.columns.contains(colOrderCol))
+    val addRowTitleDF = if(rowTitleCol != None) {df}
+    else{
+      df.withColumn("row_title", F.col(rowTitleCol.getOrElse(rowOrderCol)))
+    }
```
Comment (fucusy) [inline src/main/scala/io/github/fucusy/Viz.scala]: `val addRowTitleDF  = df.withColumn("row_title", F.col(rowTitleCol.getOrElse(rowOrderCol))`

File: src/main/scala/io/github/fucusy/Viz.scala [hunk 7]
```diff
@@ -41,26 +42,45 @@ object Viz {
...
+
+    val rowTitleColumn: Column = rowTitleCol match {
```
Comment (fucusy) [inline src/main/scala/io/github/fucusy/Viz.scala]: we don't need this, because rowTitleColumn will be always row_title

**Review by fucusy — CHANGES_REQUESTED** [review]

**Comment by fucusy** [comment]
@BAIMEI1 CI Failed

File: src/main/scala/io/github/fucusy/Viz.scala [hunk 8]
```diff
@@ -41,26 +42,42 @@ object Viz {
...
-    val rowNum = df.select("row").distinct().count().toInt
-    val tables = (1 to rowNum)
+                       rowOrderCol: String,
+                       colOrderCol: String,
+                       rowTitleCol: Option[String] = None,
+                       limitShowNumber: Int = -1
+                       ): String = {
+    require(df.columns.contains(rowOrderCol) && df.columns.contains(colOrderCol))
+    val addRowTitleDF = df.withColumn("row_title", F.col(rowTitleCol.getOrElse(rowOrderCol)).cast("string"))
+    val dropOriginRowTitleColDF = if(rowTitleCol != None){
```
Comment (fucusy) [inline src/main/scala/io/github/fucusy/Viz.scala]: rowTitleCol.isDefined

**Review by fucusy — APPROVED** [review]
LGTM

**Pull request merged**